src/replication.c

```diff
@@ -1112,15 +1113,65 @@ void restartAOFAfterSYNC() {
     }
 }
 
+long long getMonotonicTimeMs() {
+    struct timespec res;
+    clock_gettime(CLOCK_MONOTONIC, &res);
+    return (res.tv_sec * 1000000000LL + res.tv_nsec) / 1000000LL;
+}
+
+#define SLAVE_SYNC_BUDGET_UPDATE_INTERVAL_MS 10
+#define SLAVE_SYNC_BUDGET_MAX_INTERVAL_MS 50
+void updateSlaveSyncBudget(ssize_t nread) {
+    if (server.slave_sync_speed_bps <= 0) return;
+    server.slave_sync_budget -= nread * 8000LL;
+
+    long long now_ms = getMonotonicTimeMs();
+
+    long long delta_time_ms = now_ms - server.slave_sync_budget_time_ms;
+    if (delta_time_ms > SLAVE_SYNC_BUDGET_MAX_INTERVAL_MS) {
+        delta_time_ms = SLAVE_SYNC_BUDGET_MAX_INTERVAL_MS;
+    }
+    long long max_budget = SLAVE_SYNC_BUDGET_MAX_INTERVAL_MS * server.slave_sync_speed_bps;
+    server.slave_sync_budget_time_ms = now_ms;
+    server.slave_sync_budget += delta_time_ms * server.slave_sync_speed_bps;
+    if (server.slave_sync_budget > max_budget) {
+        server.slave_sync_budget = max_budget;
+    }
+}
+
+
+int slaveSyncBudgetHandler(aeEventLoop *el, long long id, void *clientData) {
+    UNUSED(clientData);
+    UNUSED(id);
+
+    updateSlaveSyncBudget(0);
+    if (server.slave_sync_budget < 0) {
+        return SLAVE_SYNC_BUDGET_UPDATE_INTERVAL_MS;
+    }
+
+    int fd = server.repl_transfer_s;
+    if (fd == -1) {
+        serverLog(LL_WARNING, "Socket is already closed, this shouldn't have happened");
+        server.slave_sync_budget_event_id = AE_DELETED_EVENT_ID;
+        return AE_NOMORE;
+    }
+    if (aeCreateFileEvent(el,fd,AE_READABLE,readSyncBulkPayload,NULL) == AE_ERR) {
+        serverLog(LL_WARNING,
+                  "Can't recreate readable event for SYNC: %s (fd=%d)",
+                  strerror(errno),fd);
+        cancelReplicationHandshake();
```

Comment:
The API changed to require `int reconnect` - should this be changed to a 1 (attempt to reconnect) (1 is the original behavior from the commit)

```
-int cancelReplicationHandshake(void) {
+int cancelReplicationHandshake(int reconnect) {
     if (server.repl_state == REPL_STATE_TRANSFER) {
         replicationAbortSyncTransfer();
         server.repl_state = REPL_STATE_CONNECT;
@@ -2468,6 +2467,16 @@ int cancelReplicationHandshake(void) {
     } else {
         return 0;
     }
+
+    if (!reconnect)
+        return 1;
+
+    /* try to re-connect without waiting for replicationCron, this is needed
+     * for the "diskless loading short read" test. */
+    serverLog(LL_NOTICE,"Reconnecting to MASTER %s:%d after failure",
+        server.masterhost, server.masterport);
+    connectWithMaster();
+
     return 1;
 }
```